Better chat history loading

cyisfor opened the pull request:
Yeah... just take these as suggestions, no guarantee I didn't break everything and just haven't noticed yet...

Commit:
Documenting our sorcery

Giving a shot at documenting what's going on and why we're doing this stuff.

Commit:
Try a few times then default to peerName

Use getChatLobbyInfo and getDistantChatStatus to get names, then use getIdDetails if those give us a real RsGxsId. Also factor the QString out to avoid needing a QString::fromUtf8 everywhere.

Commit:
Sometimes we can get a GXS ID

addChatMsg expects a RsGxsId, which doesn't work so well for chat lobbies or direct chats, or broadcasts. But if it happens to be a distant private chat, we can provide a RsGxsId for addChatMsg. Otherwise, addChatMsg silently requires an empty RsGxsId to be provided.

Commit:
Debugging

Just trying to get it to compile, fixing compiler errors, getting confused.

Commit:
History no longer laggy

Loading the chat history no longer lags my GUI, and seems to load the correct human readable names. Haven't tested the distributed chats (i.e. chat rooms) very much yet. I never use them, so I have no chat history to load.

Commit:
using a more complex state for FSM multi-stage lookups

Cleaning up that temporary cache thing, making it a more complex state object, allowing us to poll for peer details, then poll for details on that peer's profile in 2 steps.

Commit:
Rearranging logic

Moving stuff around so that it checks the state, and/or chatId at the right times.

Commit:
Profile details are in peer details

Whoops, I didn't realize you don't have to poll for profile details, since they're included in the peer details. Best leave this complex state in place though, because there's a note in libretroshare/src/retroshare/rspeers.h saying "We should definitely split this into 2 sub-structures." So polling twice might one day be needed.

Commit:
Moving state.hasName outward to avoid entering "tries" loop

Just a logical nitpick really, if state.hasName is set, all the logic for finding the name should be clearly shown as being skipped.

Commit:
Adding logic for incoming/outgoing chat

Oops, was accidentally labelling all messages as from the remote source. Needed to check historyIt->incoming to see if it should have our own name instead.

Commit:
Eliminate single use function plus union weirdness

nameForGxsId was only being used once, so inlining the code. Also a direct peer can't have a GXS Id, so maybe they should go in a union.

Commit:
Putting name in the union

Because I forgot that we refer to both .gxs and .name after the name has been calculated.

Commit:
Trying to clean up names

Already using weird labels, so may as well make sure every state change means you "have" something.

Commit:
Trying to learn C++

You can't construct a union because... or wait is it the enum you can't construct?

Commit:
lookup_state_state is dumb

I always get info structures and states mixed up...

Commit:
Finally remembered name and gxs must exist simultaneously

Removing the union stuff because C++ makes it a nightmare, and also because we can only have a union between RsGxsId and RsPeerId, but the name must be separate. Plus this is a tiny temporary lookup table, so an extra 20 or so bytes per record won't hurt anything.

Commit:
Chat history is working correctly

Fixing the last few bugs, so now it compiles. I tested it, and now all the lines from me no longer had my friend's name on them. Might test chat lobbies later.

cyisfor commented:
I just noticed that I was always loading the remote name and never the local name, so I added code to check for historyIt->incoming.

Also the force push is just because I forgot to set the author name or something.

Commit:
Got the state EXACTLY BACKWARDS

Incoming means they're the ones talking, not you. I had all the conversations precisely reversed.

Commit:
Merge branch 'master' of github.com:/RetroShare/RetroShare into better_chat_history_loading

defnax commented:
@csoler hope you dont forget this tp check

csoler commented on retroshare-gui/src/gui/chat/ChatWidget.cpp:
```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
-						{
-							std::cerr << "ChatWidget History haven't found Id Details and have wait 1 sec for it." << std::endl;
-							break;
+				const std::string
```
This is bad because std::string will truncate the data to fir first 0. You'd better use this:
std::string str = historyIt->chatPeerId.toStdString();

csoler commented on retroshare-gui/src/gui/chat/ChatWidget.cpp:
```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
+				if(!state.hasName) {
+					int tries;
+					for(tries=0;tries<4;++tries) {
+						assert(!state.hasName);
```
generally we dont use assert() except when testing the code, but not to be merged in master. assert crashes the program without warning. Here it is preferable to gracefully quit the function. Use exceptions if you prefer, or a proper return after printing an error message.

csoler commented on retroshare-gui/src/gui/chat/ChatWidget.cpp:
```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
+				const std::string
+					notbeinghashableisagoodidea((const char*)historyIt->chatPeerId.toByteArray());
+				struct lookup_info& state = statecache[notbeinghashableisagoodidea];
+				if(!state.hasName) {
```
uninitialized memory read. Consequently, this outcome of this test is random.

csoler commented on retroshare-gui/src/gui/chat/ChatWidget.cpp:
```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
+					for(tries=0;tries<4;++tries) {
+						assert(!state.hasName);
+						if(state.isDirect) {
+						GOT_DIRECT:
```
these labels are a bit misleading: we're not in a switch and there's no goto going there. If you want to mark some par of the code with some kind of milestone or so, keep this but comment out the line so we know it's not actually used.

csoler replied:
ah forget this. I didn't see the goto later on.

csoler commented on retroshare-gui/src/gui/chat/ChatWidget.cpp:
```diff
@@ -393,30 +430,99 @@ void ChatWidget::init(const ChatId &chat_id, const QString &title)
+							}
+						} else if(chatId.isLobbyId()) {
+							// XXX:
+							// no joke, this is how libretroshare/src/pqi/p3historymgr.c ACTUALLY
```
you're right. It's quite bad because it supposes that the endian will be kept between two calls. Since history is loaded and saved on the same machine, it is the case. however, that conversion is inelegant for the least.

csoler commented:
Tx Cy. Sorry for taking so long to review this. There's just 3 things to fix: proper initialization of state_cache members to avoid uninitialized memory read, remove assert(), and proper conversion of Id to string using toStdString();

Commit:
Stop RSQThreadUtils::postToObject from segfaulting internally because it's broke

Some weird synchronization bug cause postToObject to segfault, when I pass the "this" pointer to a ChatWidget, which is a valid QObject. I was posting once per history line, so by accumulating all the history lines to re-add to chat, I can turn that into a single postToObject, which doesn't trigger the segfault like doing it in a loop does.

(Maybe it's bug where you segfault if you post in thread A, to thread B, then post in thread A to thread B again before the routine in thread B finishes?)

Commit:
C++ boilerplate

C++ requires a lot of boilerplate to use structs. Pretty much just compiling, fixing "errors" and compiling again.

Commit:
More boilerplate...

I forgot this-> can't appear in those weirdo constructor intializey expression list thingies.

Commit:
std::vector fails on const

Technically I should be able to have a const QString, QDateTime etc, since C++ can definitely avoid operator= when you emplace an item into a vector. But they don't avoid it in std::vector, add needless copies of objects internally, and just fail for const members.

I'm probably doing something wrong, but whatever.

Commit:
More boilerplate

Also putting addmessages on the heap, since postToObject can't seem to figure out how to move it to the other thread.

Commit:
Merge remote-tracking branch 'origin/master' into better_chat_history_loading

Commit:
let's just assume this is always true

cyisfor commented:
Thanks for reviewing it. I definitely shouldn't have left that assert in there. As for uninitialized memory, turns out C++ makes that pretty difficult. If you read here:
https://en.cppreference.com/w/cpp/container/map/operator_at
it says that if the element does not exist, then it is inserted, and it says that inserted elements are either value-initialized, or zero-initialized. Long story short, state.hasName is always going to be false initially.

I ended up avoiding using the chat lobby ID at all, since it was supposed to be the one posting the message, not the chat lobby that got reported. So I had to mess with the code for saving chat history, so it would save that message poster, instead of telling us that every message was written by some guy named "General discussion chat (EN)" or something. So I'll push a few more commits to show what I've fixed. It's still got some issues, but... it works.

...I probably need to remove some sass from my code.
